exporters/metric/cortex/auth_test.go

```diff
@@ -153,3 +164,353 @@ func createFile(bytes []byte, filepath string) error {
 	}
 	return nil
 }
+
+// TestBuildClient checks whether the buildClient successfully creates a client that can
+// connect over TLS and has the correct remote timeout and proxy url.
+func TestBuildClient(t *testing.T) {
+	tests := []struct {
+		testName               string
+		config                 Config
+		expectedRemoteTimeout  time.Duration
+		expectedErrorSubstring string
+	}{
+		{
+			testName: "Remote Timeout with Proxy URL",
+			config: Config{
+				ProxyURL:      "123.4.5.6",
+				RemoteTimeout: 123 * time.Second,
+				TLSConfig: map[string]string{
+					"ca_file":              "./ca_cert.pem",
+					"insecure_skip_verify": "0",
+				},
+			},
+			expectedRemoteTimeout:  123 * time.Second,
+			expectedErrorSubstring: "proxyconnect tcp",
+		},
+		{
+			testName: "No Timeout or Proxy URL, InsecureSkipVerify is false",
+			config: Config{
+				TLSConfig: map[string]string{
+					"ca_file":              "./ca_cert.pem",
+					"insecure_skip_verify": "0",
+				},
+			},
+			expectedErrorSubstring: "",
+		},
+		{
+			testName: "No Timeout or Proxy URL, InsecureSkipVerify is true",
+			config: Config{
+				TLSConfig: map[string]string{
+					"ca_file":              "./ca_cert.pem",
+					"insecure_skip_verify": "1",
+				},
+			},
+			expectedErrorSubstring: "",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.testName, func(t *testing.T) {
+			// Create and start the TLS server.
+			handler := func(rw http.ResponseWriter, req *http.Request) {
+				fmt.Fprint(rw, "Successfully received HTTP request!")
+			}
+			server := httptest.NewTLSServer(http.HandlerFunc(handler))
+			defer server.Close()
+
+			// Create a certicate for the CA from the TLS server. This will be used to
+			// verify the test server by the client.
+			encodedCACert := server.TLS.Certificates[0].Certificate[0]
+			caCertPEM := pem.EncodeToMemory(&pem.Block{
+				Type:  "CERTIFICATE",
+				Bytes: encodedCACert,
+			})
+			err := createFile(caCertPEM, "./ca_cert.pem")
+			require.Nil(t, err)
+			defer os.Remove("ca_cert.pem")
+
+			// Create an Exporter client and check the timeout.
+			exporter := Exporter{
+				config: test.config,
+			}
+			client, err := exporter.buildClient()
+			require.Nil(t, err)
+			require.Equal(t, client.Timeout, test.expectedRemoteTimeout)
+
+			// Attempt to send the request and verify that the correct error occurred. If
+			// an error is expected, the test checks if the error string contains the
+			// expected error substring since the error can contain the server URL, which
+			// changes every test.
+			_, err = client.Get(server.URL)
+			if test.expectedErrorSubstring != "" {
+				require.Error(t, err)
+				errorSuffix := strings.Contains(err.Error(), test.expectedErrorSubstring)
+				require.True(t, errorSuffix)
```

Comment:
I understand that assert will stop one failure from stopping other failures from showing. What do you mean by not make fixes serialized?

Reply:
> I understand that assert will stop one failure from stopping other failures from showing. What do you mean by not make fixes serialized?

So when you use `require.*` it will cause the test to end at that point, instead of recording a test error and continuing. This means that if you have multiple errors in a test you will need to fix each serially if you use `require.*` instead of `assert.*` which will output all the errors encountered in that run.

This doesn't mean that `require.*` should never be used though. It should be used when there is no way to continue based on the measured state. Otherwise, it's best to assert things and have the user notified of all failures at once.